src/coprocessor/gadgets.rs

```diff
@@ -124,22 +134,364 @@ pub fn construct_list<F: LurkField, CS: ConstraintSystem<F>>(
+    };
+
+    let a = AllocatedPtr::alloc(&mut cs.namespace(|| "fst"), || Ok(a))?;
+    let b = AllocatedPtr::alloc(&mut cs.namespace(|| "snd"), || Ok(b))?;
+    let c = AllocatedPtr::alloc(&mut cs.namespace(|| "trd"), || Ok(c))?;
+
+    let hash = hash_poseidon(
+        &mut cs.namespace(|| "hash"),
+        vec![
+            a.tag().clone(),
+            a.hash().clone(),
+            b.tag().clone(),
+            b.hash().clone(),
+            c.tag().clone(),
+            c.hash().clone(),
+        ],
+        store.poseidon_cache.constants.c6(),
+    )?;
+
+    implies_equal(
+        &mut cs.namespace(|| "hash equality"),
+        not_dummy,
+        tuple.hash(),
+        &hash,
+    );
+
+    Ok((a, b, c))
+}
+
+/// Deconstructs an `AllocatedPtr` that's assumed to be a composition of four others
+pub fn deconstruct_tuple4<F: LurkField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    store: &Store<F>,
+    not_dummy: &Boolean,
+    tuple: &AllocatedPtr<F>,
+) -> Result<
+    (
+        AllocatedPtr<F>,
+        AllocatedPtr<F>,
+        AllocatedPtr<F>,
+        AllocatedPtr<F>,
+    ),
+    SynthesisError,
+> {
+    let (a, b, c, d) = if not_dummy.get_value() == Some(true) {
+        let idx = get_ptr(tuple, store)?.get_index4().expect("invalid Ptr");
+        let (a, b, c, d) = store.expect_4_ptrs(idx);
+        (
+            store.hash_ptr(a),
+            store.hash_ptr(b),
+            store.hash_ptr(c),
+            store.hash_ptr(d),
+        )
+    } else {
+        (ZPtr::dummy(), ZPtr::dummy(), ZPtr::dummy(), ZPtr::dummy())
+    };
+
+    let a = AllocatedPtr::alloc(&mut cs.namespace(|| "fst"), || Ok(a))?;
+    let b = AllocatedPtr::alloc(&mut cs.namespace(|| "snd"), || Ok(b))?;
+    let c = AllocatedPtr::alloc(&mut cs.namespace(|| "trd"), || Ok(c))?;
+    let d = AllocatedPtr::alloc(&mut cs.namespace(|| "frt"), || Ok(d))?;
+
+    let hash = hash_poseidon(
+        &mut cs.namespace(|| "hash"),
+        vec![
+            a.tag().clone(),
+            a.hash().clone(),
+            b.tag().clone(),
+            b.hash().clone(),
+            c.tag().clone(),
+            c.hash().clone(),
+            d.tag().clone(),
+            d.hash().clone(),
+        ],
+        store.poseidon_cache.constants.c8(),
+    )?;
+
+    implies_equal(
+        &mut cs.namespace(|| "hash equality"),
+        not_dummy,
+        tuple.hash(),
+        &hash,
+    );
+
+    Ok((a, b, c, d))
+}
+
+/// Deconstructs an `AllocatedPtr` that's assumed to encode a cons-list. It
+/// requires prior knowledge of the total number `n` of elements in the list,
+/// including a possibly terminating `nil`.
+pub fn deconstruct_list<F: LurkField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    g: &GlobalAllocator<F>,
+    not_dummy: &Boolean,
+    store: &Store<F>,
+    list: &AllocatedPtr<F>,
+    n: usize,
+) -> Result<Vec<AllocatedPtr<F>>, SynthesisError> {
+    assert!(n > 0, "list must be at least nil");
```

Comment:
It's deeply weird to me that `n` should be anything other than the list length. This seems almost impossible not to be misused at some point. Am I reading it wrong? If not, even if nothing else changes, I would change `n` to be the expected/max list length (maybe change name too), and adjust usage internally. Even if you need `let n == length - 1;`, that would address this problem.